### PR TITLE
Nonpreferred option

### PR DESCRIPTION
**Pull request by huntc**
This DeferredResult design is a simple re-factoring of CompletableFuture where by CompletableFuture's non-completion methods are removed.

Specific concerns with this interface as per CompletableFuture:
- it is not consistent with the Java 8 stream API. 
- method names such as thenApply and thenCompose are different to names used by the stream API such as map and flatMap. 
- CompletableFuture has a thenApply functor method, a thenCompose monad method, methods for handling failures, and methods for completing the future successfully or with an exception. 
- There are a number of useful helper methods that it provides, but all of these could be achieved using the methods already mentioned. 
- Methods such as thenApply have three versions: one that gets invoked synchronously by the completing thread or the calling thread if already completed; and then two “async” versions where the callbacks are invoked in the default or in a supplied execution context.
- Since CompletableFuture is not an interface it can only have one implementation. Other APIs that provide futures (Scala, Netty, Guava) cannot be wrapped by this interface, meaning that it will not play well with the rest of the ecosystem, and is therefore not a good candidate for being returned or accepted by an API.
- CompletableFuture leaks its internals including methods such as getNumberOfDependents and obtrudeValue. This leakage makes CompletableFuture appear as a low level concurrency construct and not an API abstraction that can be shared between libraries.

### CODE REVIEW

File: src/main/java/com/typesafe/scalacompat/concurrent/DeferredResult.java [hunk 1]
```diff
@@ -0,0 +1,799 @@
...
+     * exception as its cause.
+     *
+     * @return the result value
+     * @throws java.util.concurrent.CancellationException
+     *          if the computation was cancelled
+     * @throws java.util.concurrent.CompletionException
+     *          if this future completed
+     *          exceptionally or a completion computation threw an exception
+     */
+    T join();
```
Comment (viktorklang) [inline src/main/java/com/typesafe/scalacompat/concurrent/DeferredResult.java]: This shouldn't be here, it's blocking

File: src/main/java/com/typesafe/scalacompat/concurrent/DeferredResult.java [hunk 2]
```diff
@@ -0,0 +1,799 @@
...
+     *
+     * @param valueIfAbsent the value to return if not completed
+     * @return the result value, if completed, else the given valueIfAbsent
+     * @throws java.util.concurrent.CancellationException
+     *          if the computation was cancelled
+     * @throws java.util.concurrent.CompletionException
+     *          if this future completed
+     *          exceptionally or a completion computation threw an exception
+     */
+    T getNow(T valueIfAbsent);
```
Comment (viktorklang) [inline src/main/java/com/typesafe/scalacompat/concurrent/DeferredResult.java]: This shouldn't be here, it's blocking.

Comment (huntc) [inline src/main/java/com/typesafe/scalacompat/concurrent/DeferredResult.java]: It isn't blocking.

File: src/main/java/com/typesafe/scalacompat/concurrent/DeferredResult.java (hunk 1 above)
Comment (huntc) [inline src/main/java/com/typesafe/scalacompat/concurrent/DeferredResult.java]: Will be removed.

File: src/main/java/com/typesafe/scalacompat/concurrent/DeferredResult.java (hunk 2 above)
Comment (viktorklang) [inline src/main/java/com/typesafe/scalacompat/concurrent/DeferredResult.java]: You're right. I guess a better method would be `def value: Optional[T]`

Comment (huntc) [inline src/main/java/com/typesafe/scalacompat/concurrent/DeferredResult.java]: Yeah, we should probably stick to the existing interface though.

Comment (viktorklang) [inline src/main/java/com/typesafe/scalacompat/concurrent/DeferredResult.java]: Alright, deal

**Comment by viktorklang** [comment]
What I'd like to see here now in this PR is a retrofitted version of CompletableFuture implementing DeferredResult, to show that it works.

**Comment by viktorklang** [comment]
All methods returning DeferredResult in src/main/java/com/typesafe/java/util/concurrent/CompletableFuture.java should return CompletableFuture

**Comment by huntc** [comment]
No longer relevant

**Pull request closed without merging**